File: services/audit/internal/handlers/audit.go
// Package handlers provides HTTP handlers for the audit service.
package handlers

import (
	"strconv"
	"strings"
	"time"

	"github.com/gofiber/fiber/v3"

	"github.com/tenzoshare/tenzoshare/services/audit/internal/repository"
	apperrors "github.com/tenzoshare/tenzoshare/shared/pkg/errors"
)

// Handler holds the audit HTTP handlers.
type Handler struct {
	repo *repository.Repository
}

// New creates a new Handler.
func New(repo *repository.Repository) *Handler {
	return &Handler{repo: repo}
}

// ListEvents handles GET /api/v1/audit/events
// Query params: user_id (comma-separated), source (comma-separated), action, start, end, limit, offset, sort_by, sort_dir
func (h *Handler) ListEvents(c fiber.Ctx) error {
	limit := 50
	if v, err := strconv.Atoi(c.Query("limit")); err == nil && v > 0 {
		limit = v
	}
	offset := 0
	if v, err := strconv.Atoi(c.Query("offset")); err == nil && v >= 0 {
		offset = v
	}

	splitTrimmed := func(s string) []string {
		if s == "" {
			return nil
		}
		parts := strings.Split(s, ",")
		out := make([]string, 0, len(parts))
		for _, p := range parts {
			if v := strings.TrimSpace(p); v != "" {
				out = append(out, v)
			}
		}
		return out
	}

	f := repository.ListFilter{
		UserIDs:    splitTrimmed(c.Query("user_id")),
		Sources:    splitTrimmed(c.Query("source")),
		Severities: splitTrimmed(c.Query("severity")),
		Action:     c.Query("action"),
		Limit:      limit,
		Offset:     offset,
		SortBy:     c.Query("sort_by", "created_at"),
		SortDir:    c.Query("sort_dir", "desc"),
	}

	if s := c.Query("start"); s != "" {
		t, err := time.Parse(time.RFC3339, s)
		if err != nil {
			return apperrors.BadRequest("invalid start time; use RFC3339 format")
		}
		f.StartTime = &t
	}
	if s := c.Query("end"); s != "" {
		t, err := time.Parse(time.RFC3339, s)
		if err != nil {
			return apperrors.BadRequest("invalid end time; use RFC3339 format")
		}
		f.EndTime = &t
	}

	logs, total, err := h.repo.List(c.Context(), f)
	if err != nil {
		return err
	}

	return c.JSON(fiber.Map{
		"total":  total,
		"limit":  f.Limit,
		"offset": f.Offset,
		"items":  logs,
	})
}

File: web/admin-portal/src/api/audit.ts
import { request } from './client';

export interface AuditEvent {
  id: string;
  source: string;
  action: string;
  severity: 'info' | 'warning' | 'error';
  user_id: string | null;
  actor_email: string | null;
  client_ip: string | null;
  subject: string;
  payload: Record<string, unknown>;
  success: boolean;
  created_at: string;
}

export interface AuditListResponse {
  total: number;
  limit: number;
  offset: number;
  items: AuditEvent[];
}

export interface AuditFilters {
  user_ids?: string[];
  sources?: string[];
  severities?: string[];
  action?: string;
  start?: string;
  end?: string;
  limit?: number;
  offset?: number;
  sort_by?: string;
  sort_dir?: string;
}

export async function listAuditEvents(filters: AuditFilters = {}): Promise<AuditListResponse> {
  const params = new URLSearchParams();
  if (filters.user_ids?.length) params.set('user_id', filters.user_ids.join(','));
  if (filters.sources?.length) params.set('source', filters.sources.join(','));
  if (filters.severities?.length) params.set('severity', filters.severities.join(','));
  if (filters.action) params.set('action', filters.action);
  if (filters.start) params.set('start', filters.start);
  if (filters.end) params.set('end', filters.end);
  if (filters.sort_by) params.set('sort_by', filters.sort_by);
  if (filters.sort_dir) params.set('sort_dir', filters.sort_dir);
  params.set('limit', String(filters.limit ?? 50));
  params.set('offset', String(filters.offset ?? 0));

  return request<AuditListResponse>(`/audit/events?${params.toString()}`);
}
